### ISSUE TITLE
'cabal repl' with stack traces

### ISSUE THREAD

**Issue by emlautarom1** [haskell/cabal#7674]
I've been looking around for a set of flags/configurations that would allow me to load my project into GHCi and have access to stack traces. So far, I found [this blog post](http://simonmar.github.io/posts/2016-02-12-Stack-traces-in-GHCi.html) that explains how to access stack traces from plain GHCi (not from `cabal repl`). I also found [this](http://downloads.haskell.org/~ghc/latest/docs/html/users_guide/ghci.html#stack-traces-in-ghci) in the official docs - which is practically a summary of the blog post.

What I've tried so far with no success:

- Add `-fexternal-interpreter -prof` to the `ghc-options` section in my `.cabal file`
- Run `cabal repl --ghc-options="-fexternal-interpreter -prof"`. I got a warning from GHCi, but no stack traces.
- Run `cabal repl -p`. This required installing the `ghc-prof` library (I'm using Fedora).
- Run `cabal exec ghci -- -fexternal-interpreter -prof`. I got some errors related to missing files in the `new-dist` folder. Then, I recompiled with `cabal build -p`.

My project has several targets (a lib, a test suite and several executables). What I want is to load the library (the default when running `cabal repl`) and use `Debug.traceStack`. I do not care about getting stack traces from third party dependencies, I just care about what my code is doing.

Right now I'm using `HasCallStack` as a workaround, but the traces are extremely short (a single function) and it's very inconvenient to add the constraint on every function.

**Comment by fgaz** [comment]
What were the results of the last two bullet points? `-p` should work iirc

**Comment by emlautarom1** [comment]
Minimal working example: run `cabal init --lib`. Edit `MyLib.hs` with the following code:

```haskell
module MyLib (someFunc) where

import qualified Debug.Trace as Debug

f :: Int -> Int
f x = Debug.traceStack "> from f" (x + 1)

someFunc :: IO ()
someFunc = do
  let x = 5
  let res = f x
  putStrLn "result"
  print res
  putStrLn "done"
```

Start REPL using `cabal repl -p`. Evaluate `someFunc`. Output is:

```
*MyLib> someFunc 
result
> from f
6
done
```

I expect to get a stack trace, yet I get none. It behaves like `Debug.trace`.

**Comment by emlautarom1** [comment]
I'm using cabal 3.0 by the way (`cabal-install version 3.0.0.0`; `compiled using version 3.0.1.0 of the Cabal library`)

**Comment by fgaz** [comment]
```
cabal build -p
cabal exec ghci -- -fexternal-interpreter -prof
```

works for me. I'm using master, but I'd espect 3.6 to work too (3.0 is quite old by the way!)

Still, I think profiling flags should be passed to ghci in cabal repl as well. The one above is just a workaround.

**Comment by emlautarom1** [comment]
Indeed, that works. I need to load the library once in GHCi though (`:l MyLib`) though.
Sadly this does not work on my main project which uses several third party dependencies. I get errors related to shared objects (after running `cabal exec ghci -- -fexternal-interpreter -prof`):

```
GHCi, version 8.8.4: https://www.haskell.org/ghc/  :? for help
Loaded package environment from /home/emlautarom1/Development/Haskell/duckling/dist-newstyle/tmp/environment.-176033/.ghc.environment.x86_64-linux-8.8.4
Loaded package environment from /home/emlautarom1/Development/Haskell/duckling/dist-newstyle/tmp/environment.-176033/.ghc.environment.x86_64-linux-8.8.4
<command line>: libHSextensible-exceptions-0.1.1.4-66ee67394de71dfcdfc753af01d17a538e5fbe79e155b15ebbc55afc2a4fffe8.so: cannot open shared object file: No such file or directory
```

I found that I could fix this issue by installing `ghc-extensible-exceptions` (I'm using Fedora btw). But then there's another issue:

```
Resolving dependencies...
GHCi, version 8.8.4: https://www.haskell.org/ghc/  :? for help
Loaded package environment from /home/emlautarom1/Development/Haskell/duckling/dist-newstyle/tmp/environment.-176275/.ghc.environment.x86_64-linux-8.8.4
Loaded package environment from /home/emlautarom1/Development/Haskell/duckling/dist-newstyle/tmp/environment.-176275/.ghc.environment.x86_64-linux-8.8.4
<command line>: libHStimezone-series-0.1.9-1ee495d5ea91c01050b65c31aa846fca19aa17022b9c45c7ed21d61bb40e17d8.so: cannot open shared object file: No such file or directory
```

So, again, I need to install `ghc-timezone-series`. I guess that I would need to do that for every dependency, which seems totally unreasonable. Could be the fact taht I installed GHC through my package manager (`dnf`) and not `stack` or `ghcup` the cause of those missing shared objects?

**Comment by fgaz** [comment]
Is that with cabal-install 3.6?

**Comment by emlautarom1** [comment]
I'm using cabal 3.0, plain commands which default to `v2` versions. I haven't tried it yet with newer versions since they're not available through `dnf`. I could try using the latest version just to ensure that it works as expected now.

**Comment by Mikolaj** [comment]
With some luck `cabal install cabal-install` should bring you new cabal, so you don't have to use any special tool for that.
